### ISSUE TITLE
Multi-store cache conflict with Varnish

### ISSUE THREAD

**Issue by br33f** [magento/magento2#38483]
### Preconditions and environment
Magento 2.4.3-p2 and 2.4.6 for sure. Perhaps other versions too.

1. There must be at least two store views configured.
![image](https://github.com/magento/magento2/assets/14214995/fe1b42ea-56b5-4f01-8e14-5fe7013b165e)
2. Multi-store must be set up on a single domain (using cookie)
3. Full Page Cache > Caching application should be set to Varnish.
![image](https://github.com/magento/magento2/assets/14214995/dd299d53-8e02-4549-aa5b-fea94eeab95e)
4. Varnish configuration is set according to the configuration generated with "Export VCL for Varnish 6"

### Steps to reproduce

1. Enter a page (let's call it X) that is available on both store views (and URL is the same)
2. Switch store to non-default (let's call it B)
3. Clear all cache (including Varnish)
4. Go to Developer tools and delete "X-Magento-Vary" cookie. Leave other cookies (store etc.) as it is.
5. Refresh page
6. Try to switch back to default store view (let's call it A)

### Expected result

Store view switches back to default store view. The page shows content from store view A (default).

### Actual result

You can not switch store view back to default store view on the page you visited. 
It works fine on another pages. However this particular page X is being returned from Varnish which (because the lack of X-Magento-Vary cookie) cached store B contents as default (A) (without X-Magento-Vary cookie).

Even if you switch store view to default on another page and then navigate to Page X - store switcher will show invalid store and page content is still wrong.

### Additional information

Varnish stores cache according to X-Magento-Vary cookie. If this cookie is missing - Varnish assumes that it is a default store view. However one can make a request without X-Magento-Vary cookie, but still with "store" cookie.
Magento does not validate that, it returns content for specified "store" - but Varnish caches it without X-Magento-Vary.

Then if you visit page with default store view (X-Magento-Vary cookie not present) - Varnish serves you with invalid store view page.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by br33f** [comment]
I don't think it is reproducible on the Magento instance that magento bot gives you. I don't know if it is possible to configure Varnish properly on that instance - the default is Built-in cache.

**Comment by andreifurrnica** [comment]
@magento give me 2.4.6-p3 instance

**Comment by andreifurrnica** [comment]
@magento give me 2.4.6-p3 instance

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @br33f,

Thank you for reporting and collaboration.

Verified the issue on Magento 2.4-develop instance and the issue is reproducible.Kindly refer the screenshots.

**Steps to reproduce**

- Enter a page (let's call it X) that is available on both store views (and URL is the same)
- Switch store to non-default (let's call it B)
- Clear all cache (including Varnish)
- Go to Developer tools and delete "X-Magento-Vary" cookie. Leave other cookies (store etc.) as it is.
- Refresh page
- Try to switch back to default store view (let's call it A)



![Screenshot from 2024-03-08 13-05-59](https://github.com/magento/magento2/assets/51680745/1c5785da-04af-4d2f-ba74-e3cd27162c57)

We are not able to switch to default store view.

Hence Confirming the issue.

Thanks.

**Comment by ln8711** [comment]
Seem like function `getCurrentStoreId` from `vendor/magento/module-store/Model/StoreResolver.php` doesn't get correct store code at this part:
    $storeCode = $this->request->getParam(
                \Magento\Store\Model\StoreManagerInterface::PARAM_NAME,
                $this->storeCookieManager->getStoreCodeFromCookie()
     );
 so the result is it returns the default store instead.
 
 I had to add some code after that to detect the store view using HTTP_HOST value by comparing it with all stores base url (except admin store) to return correct store code after the line above.

**Comment by Nico10** [comment]
any fix for this ?

**Comment by VasilisX** [comment]
The same issue for magento 2.4.6-p3 for me too. Is there any workaround?

**Comment by lan-n-ttm** [comment]
I created a work around by add a function to detect store code base on request (using the requested HTTP_HOST variable)
in file  `vendor/magento/module-store/Model/StoreResolver.php`
And seem it works for my case


```
      if (!$storeCode) {
            $storeCode = $this->request->getParam(
                \Magento\Store\Model\StoreManagerInterface::PARAM_NAME,
                $this->storeCookieManager->getStoreCodeFromCookie()
            );

            if (!$storeCode) {
                /** add a function to detect the store code by HTTP_HOST */
                $storeCodeFromReqest = $this->getStoreCodeByRequest();
                if ($storeCodeFromReqest) {
                    $storeCode = $storeCodeFromReqest;
                }
            }
        }
```

The function getStoreCodeByRequest just takes the HTTP_HOST and compare with all store URLs to find a matched one and return the store code.

**Comment by lbajsarowicz** [comment]
This multi-store Varnish cache-key issue is tied to a missing X-Magento-Vary cookie versus a store-cookie mismatch; it requires a live Varnish plus two-store setup to verify and can't be confirmed from static code. It's related to but distinct from #37808 (both Varnish/multi-store caching issues but different symptoms). Please re-QA with a two-store Varnish setup on current 2.4-develop to confirm and isolate the cache-key mismatch.
